pkg/chunk/elastic/es_index_client.go

```diff
@@ -0,0 +1,344 @@
+	"time"
+
+	"reflect"
+
+	"fmt"
+	"github.com/olivere/elastic"
+)
+
+const (
+	null         = string('\xff')
+	maxFetchDocs = 1000
+)
+
+// Config for a BoltDB index client.
+type Config struct {
+	Address       string `yaml:"address"`
+	IndexType     string `yaml:"index_type"`
+	User          string `yaml:"user"`
+	Password      string `yaml:"password"`
+	TLSSkipVerify bool   `yaml:"tls_skip_verify"`
+	CertFile      string `yaml:"cert_file"`
+	KeyFile       string `yaml:"key_file"`
+	CaFile        string `yaml:"ca_file"`
+}
+
+// RegisterFlags registers flags.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.Address, "elastic.address", "http://127.0.0.1:9200", "Address of ElasticSearch.")
+	f.StringVar(&cfg.IndexType, "elastic.index_type", "lokiindex", "Index Type used in ElasticSearch.")
+	f.StringVar(&cfg.User, "elastic.user", "user", "User used in ElasticSearch Basic Auth.")
+	f.StringVar(&cfg.Password, "elastic.password", "password", "Password used in ElasticSearch Basic Auth.")
+	f.BoolVar(&cfg.TLSSkipVerify, "elastic.tls_skip_verify", true, "Skip tls verify or not. Default is skip.")
+	f.StringVar(&cfg.CertFile, "elastic.cert_file", "", "Cert File Location used in TLS Verify.")
+	f.StringVar(&cfg.KeyFile, "elastic.key_file", "", "Key File Location used in TLS Verify.")
+	f.StringVar(&cfg.CaFile, "elastic.ca_file", "", "CA File Location used in TLS Verify.")
+
+}
+
+// IndexEntry describes an entry in the chunk index
+type IndexEntry struct {
+	Hash  string `json:"hash"`
+	Range string `json:"range"`
+	Value string `json:"value,omitempty"`
+}
+
+const mapping = `
+{
+	"settings":{
+		"number_of_shards": 1,
+		"number_of_replicas": 0
+	},
+	"mappings":{
+		"lokiindex":{
+			"properties":{
+				"hash":{
+					"type":"keyword"
+				},
+				"range":{
+					"type":"keyword"
+				},
+				"value":{
+					"type":"keyword"
+				}
+			}
+		}
+	}
+}`
+
+var client *elastic.Client
+
+// Starting with elastic.v5, you must pass a context to execute each service
+var ctx = context.Background()
+
+// StorageClient implements chunk.IndexClient and chunk.ObjectClient for Cassandra.
+type esClient struct {
+	cfg    Config
+	client *elastic.Client
+}
+
+func (e *esClient) Stop() {
+	e.client.Stop()
+}
+
+// ES batching isn't really useful in this case, its more to do multiple
+// atomic writes.  Therefore we just do a bunch of writes in parallel.
+type writeBatch struct {
+	entries []chunk.IndexEntry
+}
+
+func (e *esClient) NewWriteBatch() chunk.WriteBatch {
+	return &writeBatch{}
+}
+
+func (b *writeBatch) Add(tableName, hashValue string, rangeValue []byte, value []byte) {
+	b.entries = append(b.entries, chunk.IndexEntry{
+		TableName:  tableName,
+		HashValue:  hashValue,
+		RangeValue: rangeValue,
+		Value:      value,
+	})
+}
+
+func (e *esClient) BatchWrite(ctx context.Context, batch chunk.WriteBatch) error {
+	b := batch.(*writeBatch)
+
+	indexName := b.entries[0].TableName
+	exists, err := e.client.IndexExists(indexName).Do(ctx)
```

Comment:
A check for an index on every write seems inefficient in the pattern that cortex uses. I'm only starting to work through the review on this, however I know that Elasticsearch has an auto-create index feature and templates. This can be used to create the index if it doesn't already exist on the first bulk index request, plus having the mapping template applied automatically. 

The management of the template(s) would be something I would expect the tablemanager to do.

Reply:
you're right. I will change it.